### ISSUE TITLE
wrong semantics on lookup vindex for insert ignore and upsert queries

### ISSUE THREAD

**Issue by demmer** [youtube/vitess#3012]
When handling `INSERT IGNORE` or `INSERT ... ON DUPLICATE KEY UPDATE ...` queries on a table with a lookup vindex, vtgate fails the operation if an entry already exists in the secondary index table.

Instead I think the right behavior should be:
* For `INSERT IGNORE` it should succeed even if the secondary vindex query fails for any reason.
* For upsert queries where the fields in the secondary index _are not_ included in the update clause, it should try the insert but ignore errors if the key already existing in the table.
* For upsert queries where the fields in the secondary index _are_ included in the update clause, it should probably pass through a sensible update statement but this also seems quite unlikely so IMO it seems ok to just fail to handle this case for now.

**Comment by sougou** [comment]
## Design notes

There are actually three possible combinations beyond the simple `INSERT`:
1. `INSERT IGNORE`: This statement ignores 'all' errors including dup keys, which is usually not what the app intends.
2. `INSERT...ON DUPLICATE KEY`: This statement performs an update if there is duplicate key, but returns an error if the update encounters dup keys.
3. `INSERT IGNORE... ON DUPLICATE KEY`: This statement attempts to insert. On dup key, it attempts to update. If the update has dup key errors, then it ignores the failures.

We can disallow construct 3. There shouldn't be many use cases for it, and app-level workarounds can be coded if such a need really arises.

*CAVEAT: The solution below opens a hole in the owned vindex design: It will allow the possibility of multiple owner rows for a single vindex entry. If this happens, deleting one of the owner rows will render the other owner orphaned. The use cases that are requesting this feature will generally not do this. But it's a caveat to keep in mind. The reason why we should still allow this is because the lookup vindex is mainly a convenience feature that allows you to simplify the app. It's not meant to be an integrity enforcer in the relational sense.*

### INSERT IGNORE
For the first construct, the following changes need to be made:
1. For unowned vindexes with values supplied: we perform a sanity check that the supplied values map to the target keyspace id. With the `ignore` construct, we can return success with 0 rows affected.
2. For unowned vindexes with no values supplied: we try to compute the value if a reverse map is available. If this fails, we can return success with 0 rows affected.
3. For owned vindexes: We try to create an entry. If this fails, then the code needs to switch to behavior 1.

The additional complexity is that the inserts may have multiple values. If so, the rules should be individually applied for each row, but the whole operation still needs to be atomic. For example, the first row may need a vindex row to be created, but the second row may see a duplicate, but the sanity check could fail. If so, the previously created vindex must be rolled back. This can be achieved with the following steps:
1. Perform the sanity checks for rows that are already present.
2. Compute which rows are absent (may be derive it as side-effect of step 1).
3. Insert just entries for the rows that are absent.

If any failure is encountered while performing the above, vtgate can fail the entire `insert ignore` and return 0 rows updated.

As a future improvement, we could remove the insert entries that generated errors and just send the ones that succeeded.

Additional possible improvement: MySQL ignores all errors on `INSERT IGNORE`. We could instead ignore only data errors, but return failure on other system-level errors.

### INSERT...ON DUPLICATE KEY
For this construct, we do everything that `INSERT IGNORE` performs. However, we additionally have to verify that the the dup key statements don't update the vindex columns. This rule is already enforced for all updates, and should be enforced the same way. This will be detected at the time of analysis and the failure will be before any statements are executed.

Possible improvement: For this construct, we should only ignore data errors. Other system errors will need to be returned back to the client.

**Comment by sougou** [comment]
There are a few more corner cases to worry about:
* After a successful insert in the lookup, the main insert can fail for reasons other than dup key. If so, no error will be returned, but the lookup entry will survive without an owner row. This will cause subsequent regular inserts to chronically fail. However, an `INSERT IGNORE` that succeeds will fix this inconsistency. The other way out is to manually drop the row.
* Behavior for unique and non-unique lookup may have to diverge: In the case of a unique lookup, a key being present but the value not matching means that we want to skip the insert. In the case of a non-unique lookup, we need to create a new association and proceed with the insert.

This further encourages us to redefine the meaning of `INSERT IGNORE` to return an error if the failure is anything other than a dup key error. If this happened, the distributed transaction watcher can detect partial executions and abort such transactions, thereby preventing dangling rows.

**Comment by sougou** [comment]
After some more study, I think an UPSERT would be a challenge to implement in a 100% MySQL compatible way. But we can come up with a decent trade-off that will work as intended for practical use cases:
1. For a table that owns a vindex, the upsert should try to create an entry for the lookup vindex. On duplicate key, we have to verify that there is no routing conflict: there must exist a route from the vindex key to the keyspace id of the row. If such a route does not exist, then the request is inconsistent, and we must not route that part of the request. This means that no rows will be affected. Contrasting this with MySQL, this is equivalent to the case where a table has multiple unique keys; MySQL will choose one of those rows and update it.
2. As mentioned previously, if an entry already exists for an owned vindex, and the keyespace id matches, then we can route that statement to the correct shard. However, it is possible that the UPSERT would eventually insert a row instead of updating one, which will cause two rows to own a vindex entry. We've decided that allowing this is an acceptable trade-off.
3. For an unowned vindex, if the keyspace id doesn't match, we'll return success with 0 rows affected, just like case 1. If the keyspace id matches, then we route the query to that shard.

Simply put, a row can have only one keyspace id. If an operation can result in this rule being broken, upserts will not be forwarded and will instead return 0 rows affected. Should we instead return an error?

I'll post a separate brainstorm on how to handle multi-value upserts on the vttablet side.

**Comment by sougou** [comment]
For the vttablet I continued @demmer's previous work:
* If a table has only one unique key, then the query is passed through. Also, multiple rows are allowed.
* If the updates modify any of the primary key columns, then those values are also added to the stream comment.
* Rewrote `VALUES(column)` handling to compute and print out row-wise values for multi-row upserts.
* In the case where a table has more than one unique key, we continue with the plan that breaks it up into two statements. We allow only one row in such cases. Previously, the generated update used a `Resolved` field, but the new implementation uses a custom formatter instead. Changes can be previewed here https://github.com/youtube/vitess/pull/3042.

**Comment by sougou** [comment]
To implement the functionality on the VTGate side, the vindex API needs to be extended: A lookup vindex will need to support a `CreateIgnore` function. This function will create what it can and will not return errors on dup key.

The `Verify` API will need to be changed to return a `[]bool` (instead of a single `bool`), one for each association. This will allow VTGate to decide which rows must be dropped.

While inspecting lookup, I found a minor bug in its Verify function that would have caused it to falsely fail some multi-row inserts: If identical pairs were sent as input, the number of results returned will be less than the input length (because two identical conditions matched still returns only one row), which would cause it to think that everything didn't verify. This is a very uncommon use case, which is why no one has seen it yet.

This can be fixed one of two ways:
1. Perform the verification one row at a time. This will be the most correct, but could slow down multi-row inserts.
2. Pull all results at once, build a map and performs a hash comparison. This will be fast. However, hash comparison is lexical and will not work correctly for text columns.

For now, I'll go with option 1 for the sake of correctness. There are ways to address issues raised in option 2, but the solutions are more elaborate. So, we'll have to tackle that separately.
